frescobaldi_app/sessions/menu.py

```diff
@@ -35,43 +35,71 @@ class SessionMenu(QMenu):
     def __init__(self, mainwindow):
         super(SessionMenu, self).__init__(mainwindow)
         app.translateUI(self)
-        mgr = manager.get(mainwindow)
-        ac = mgr.actionCollection
-        ag = self._actionGroup = QActionGroup(self)
-        ag.setExclusive(True)
-        ag.addAction(ac.session_none)
-        ag.triggered.connect(self.slotSessionsAction)
-        self.addAction(ac.session_new)
-        self.addAction(ac.session_save)
-        self.addSeparator()
-        self.addAction(ac.session_manage)
-        self.addSeparator()
-        self.addAction(ac.session_none)
-        self.addSeparator()
         self.aboutToShow.connect(self.populate)
 
     def translateUI(self):
         self.setTitle(_('menu title', '&Session'))
 
     def populate(self):
-        ac = manager.get(self.parentWidget()).actionCollection
-        ag = self._actionGroup
-        for a in ag.actions():
-            if a is not ac.session_none:
-                self.removeAction(a)
-                ag.removeAction(a)
-        ac.session_none.setChecked(not sessions.currentSession())
-        for name in sessions.sessionNames():
-            a = self.addAction(name.replace('&', '&&'))
+        groups = {}
+        tl_sessions = []
+
+        def add_session(menu, name, group=None):
+            name = name.replace('&', '&&')
+            fullname = group + '/' + name if group else name
+            a = menu.addAction(name)
             a.setCheckable(True)
-            if name == sessions.currentSession():
+            if fullname == sessions.currentSession():
                 a.setChecked(True)
-            a.setObjectName(name)
-            ag.addAction(a)
+                if group:
+                    menu.setTitle('* ' + menu.title())
+            a.setObjectName(fullname)
+            self._actionGroup.addAction(a)
+
+        def add_sessions():
+            for name in tl_sessions:
+                add_session(self, name)
+
+        def add_groups():
+            g_keys = [k for k in groups.keys()]
+            g_keys.sort()
+            for k in sorted(groups.keys()):
+                m = self.addMenu(k)
+                for name in groups[k]:
+                    add_session(m, name, k)
+                qutil.addAccelerators(m.actions())
+
+        def reset_menu():
+            self.clear()
+            ac = manager.get(self.parentWidget()).actionCollection
+            ac.session_none.setChecked(not sessions.currentSession())
+            ag = self._actionGroup = QActionGroup(self)
+            ag.setExclusive(True)
+            ag.addAction(ac.session_none)
+            ag.triggered.connect(self.slotSessionsAction)
+            self.addAction(ac.session_new)
```

Comment:
Do we need to add these each time now? Previously they were just added on init.

Reply:
I may be wrong (would be happy) but after quite some searching I didn't find a way to *remove* a submenu from a `QMenu`. But it is important to remove the submenus each time the menu is populated, so I found it's the easiest way to just clear the whole menu. After all these actions aren't *created* but only *added*. 
But I'd be happy to learn about a cleaner way to achieve it.

Reply:
I can dig deeper into this if you like. But as I understand it you remove (also) submenus with `removeAction`. It's in the original code.

Reply:
Unfortunately it doesn't work like that:

![image](https://user-images.githubusercontent.com/1812148/37592390-cb3a9aae-2b6e-11e8-939e-dc94b88d7b24.png)

I'll ask on stackoverflow if it's possible to remove submenus.

Reply:
Sorry, should have looked in *that* direction before sending the previous comment:

It's possible to remove the submenu with `<submenu>.deleteLater()`. I'll update the code and push something soon.

Reply:
Ok, I guess that resolves the issues I had initially!